kedro/io/kedro_data_catalog.py

```diff
@@ -0,0 +1,326 @@
+        return copy.copy(self._datasets)
+
+    @datasets.setter
+    def datasets(self, value: Any) -> None:
+        raise AttributeError(
+            "Operation not allowed! Please use KedroDataCatalog.add() instead."
+        )
+
+    @property
+    def config_resolver(self) -> CatalogConfigResolver:
+        return self._config_resolver
+
+    def __repr__(self) -> str:
+        return self._datasets.__repr__()
+
+    def __contains__(self, dataset_name: str) -> bool:
+        """Check if an item is in the catalog as a materialised dataset or pattern"""
+        return (
+            dataset_name in self._datasets
+            or self._config_resolver.match_pattern(dataset_name) is not None
+        )
+
+    def __eq__(self, other) -> bool:  # type: ignore[no-untyped-def]
+        return (self._datasets, self._config_resolver.list_patterns()) == (
+            other._datasets,
+            other.config_resolver.list_patterns(),
+        )
+
+    @property
+    def _logger(self) -> logging.Logger:
+        return logging.getLogger(__name__)
+
+    @classmethod
+    def from_config(
+        cls,
+        catalog: dict[str, dict[str, Any]] | None,
+        credentials: dict[str, dict[str, Any]] | None = None,
+        load_versions: dict[str, str] | None = None,
+        save_version: str | None = None,
+    ) -> KedroDataCatalog:
+        """Create a ``KedroDataCatalog`` instance from configuration. This is a
+        factory method used to provide developers with a way to instantiate
+        ``KedroDataCatalog`` with configuration parsed from configuration files.
+        """
+        catalog = catalog or {}
+        config_resolver = CatalogConfigResolver(catalog, credentials)
+        save_version = save_version or generate_timestamp()
+        load_versions = load_versions or {}
+
+        missing_keys = [
+            ds_name
+            for ds_name in load_versions
+            if not (
+                ds_name in config_resolver.config
+                or config_resolver.match_pattern(ds_name)
+            )
+        ]
+        if missing_keys:
+            raise DatasetNotFoundError(
+                f"'load_versions' keys [{', '.join(sorted(missing_keys))}] "
+                f"are not found in the catalog."
+            )
+
+        return cls(
+            load_versions=load_versions,
+            save_version=save_version,
+            config_resolver=config_resolver,
+        )
+
+    @staticmethod
+    def _validate_dataset_config(ds_name: str, ds_config: Any) -> None:
+        if not isinstance(ds_config, dict):
+            raise DatasetError(
+                f"Catalog entry '{ds_name}' is not a valid dataset configuration. "
+                "\nHint: If this catalog entry is intended for variable interpolation, "
+                "make sure that the key is preceded by an underscore."
+            )
+
+    def _add_from_config(self, ds_name: str, ds_config: dict[str, Any]) -> None:
```

Comment:
Perhaps name this `_add_dataset_from_config` to make it specific that this is calling the `AbstractDataset.from_config`

Reply:
Since we have `add()` method and `add_all()` (for `DataCatalog`) that add datasets without specifying that in their names I ended with the current naming. We only specify the type of data for `add_feed_dict` / `add_raw_data`.

I don't insist on the current naming though but it looks more aligned with what we already have.